### PR TITLE
ft(S3C-3707): Add disk usage soft limit task

### PR DESCRIPTION
**Pull request by tmacro**

### CODE REVIEW

**Comment by vrancurel** [comment]
Can we have a unit or functional test ?

**Comment by tmacro** [comment]
@vrancurel yes, I'm working on some now

File: libV2/config/index.js [hunk 1]
```diff
@@ -297,8 +297,27 @@ class Config {
...
+                _typeCasts.int,
+            ),
         };
 
+
+        if (!diskUsage.path && (diskUsage.softLimit !== undefined || diskUsage.hardLimit !== undefined)) {
+            throw Error('You must specify diskUsage.path to monitor for disk usage');
+        } else if (diskUsage.path && (diskUsage.softLimit === undefined && diskUsage.hardLimit === undefined)) {
+            throw Error('One of diskUsage.softLimit or diskUsage.hardLimit must be specified');
+        }
```
Comment (miniscruff) [inline libV2/config/index.js:315]: I think the if checks here are a little redundant if I am understanding it right. Is there a reason why this wouldn't work?

```suggestion
        if (!diskUsage.path) {
            throw Error('You must specify diskUsage.path to monitor for disk usage');
        }
        
        if (diskUsage.softLimit === undefined && diskUsage.hardLimit === undefined) {
            throw Error('One of diskUsage.softLimit or diskUsage.hardLimit must be specified');
        }
```

File: tests/functional/softLimit/testSoftLimit.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+    it('should trigger delete if above the limit', async () => {
+        fillDir(path, { count: 1, size: 100 });
+        task._softLimit = 1;
+        const checkSpy = sinon.spy(task, '_checkSoftLimit');
+        const expireSpy = sinon.spy(task, '_expireMetrics');
+        const deleteStub = sinon.spy(task._warp10, 'delete');
+        await writeEvent(task);
+        await writeEvent(task);
+        // await sleep(1000);
```
Comment (miniscruff) [inline tests/functional/softLimit/testSoftLimit.js]: commented line

File: libV2/config/index.js (hunk 1 above)
Comment (tmacro) [reply libV2/config/index.js:315]: The logic is meant to be, if you define `path` then you must define `softLimit` or `hardLimit`and conversely if you define `softLimit` or `hardLimit` you must define `path`.
Defining nothing is also an option as this config file is shared amonst the various utapi components which might not need the `diskUsage` section

Comment (tmacro) [reply libV2/config/index.js:315]: Also we need this checking here as config loading can happen completely though the environment

Comment (miniscruff) [reply libV2/config/index.js:315]: oooh ok, that is more complex than I was thinking. It looks good then, that matches.

File: tests/functional/softLimit/testSoftLimit.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
+const assert = require('assert');
+const fs = require('fs');
+const sinon = require('sinon');
+const uuid = require('uuid');
+
+const { MonitorDiskUsage } = require('../../../libV2/tasks');
+const { UtapiMetric } = require('../../../libV2/models');
+const { now } = require('../../../libV2/utils');
+
+function fillDir(path, { count, size }) {
```
Comment (miniscruff) [inline tests/functional/softLimit/testSoftLimit.js]: not sure the need of destructuring count and size here

File: tests/functional/softLimit/testSoftLimit.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        new UtapiMetric({ timestamp: now(), bucket: 'foo' }),
+    ]);
+    return task._warp10.ingest({ className: 'utapi.event' }, [
+        new UtapiMetric({ timestamp: now(), bucket: 'foo' }),
+    ]);
+}
+
+// eslint-disable-next-line func-names
+describe('Test MonitorDiskUsage soft limit', function () {
+    this.timeout(30000);
```
Comment (miniscruff) [inline tests/functional/softLimit/testSoftLimit.js:22]: does this set the timeout for all the tests? Never seen it before.

File: tests/functional/softLimit/testSoftLimit.js [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    });
+
+    it('should trigger delete if above the limit', async () => {
+        fillDir(path, { count: 1, size: 100 });
+        task._softLimit = 1;
+        const checkSpy = sinon.spy(task, '_checkSoftLimit');
+        const expireSpy = sinon.spy(task, '_expireMetrics');
+        const deleteStub = sinon.spy(task._warp10, 'delete');
+        await writeEvent(task);
+        await writeEvent(task);
```
Comment (miniscruff) [inline tests/functional/softLimit/testSoftLimit.js]: just want to make sure you meant to call this twice, it doesn't seem to me to be required.

File: libV2/tasks/DiskUsage.js [hunk 6]
```diff
@@ -16,29 +25,147 @@ class MonitorDiskUsage extends BaseTask {
         this._defaultLag = 0;
         this._path = config.diskUsage.path;
         this._enabled = config.diskUsage.enabled;
+        this._mode = config.diskUsage.mode;
```
Comment (miniscruff) [inline libV2/tasks/DiskUsage.js:28]: Not sure if this is required but I do not see `mode` or `enabled` in the config schema.

File: libV2/tasks/DiskUsage.js [hunk 7]
```diff
@@ -16,29 +25,147 @@ class MonitorDiskUsage extends BaseTask {
...
+                            moduleLogger.warn('no disk usage entries found');
+                        } else {
+                            shouldDelete = JSON.parse(resp.result)
+                                .map(val => ({ node: val.l.node, used: val.v[0][1] }))
+                                .map(val => this._checkSoftLimit(val.used, val.node))
+                                .some(val => val);
+                        }
+                    } catch (error) {
+                        moduleLogger.error('error fetching disk usage data from warp 10', { error });
+                    }
```
Comment (miniscruff) [inline libV2/tasks/DiskUsage.js:147]: not sure how feasible it is but neither catch in this method has tests, or at least I am guessing so as I do not see tests with `assert.throws`.

File: tests/functional/softLimit/testSoftLimit.js (hunk 3 above)
Comment (tmacro) [reply tests/functional/softLimit/testSoftLimit.js]: This is a copy/paste of a helper function in a different suite thus the needless destructuring. I'll pull it out to the test utils

File: tests/functional/softLimit/testSoftLimit.js (hunk 4 above)
Comment (tmacro) [reply tests/functional/softLimit/testSoftLimit.js:22]: It does. Only for the tests under the `describe`

File: tests/functional/softLimit/testSoftLimit.js (hunk 5 above)
Comment (tmacro) [reply tests/functional/softLimit/testSoftLimit.js]: left over from iding a bug

File: libV2/tasks/DiskUsage.js (hunk 7 above)
Comment (tmacro) [reply libV2/tasks/DiskUsage.js:147]: I can stub it with `sinon`

File: libV2/tasks/DiskUsage.js (hunk 6 above)
Comment (miniscruff) [reply libV2/tasks/DiskUsage.js:28]: what about diskUsage.enabled?

**Review by miniscruff — APPROVED** [review]

**Comment by tmacro** [comment]
@miniscruff `diskUsage.enabled` is a computed property not found in the user config

**Review by anurag4DSB — APPROVED** [review]

**Pull request merged**